### ISSUE TITLE
Review Extensions conformance section

### ISSUE THREAD

**Issue by jpmckinney** [open-contracting/standard#582]
Possible new requirements for extensions:

* All definitions and properties must set a `title`, `description` and `type`, unless they are originally defined in the core schema or in another extension. https://github.com/open-contracting/standard/issues/603
* If `openCodelist` is `true`, `enum` must not be set.
* If `openCodelist` is `false`, `enum` must be set. https://github.com/open-contracting/ocds-extensions/issues/32
* If `openCodelist` is `false` and `enum` is set, `enum` must match the codes in the `codelist`.
* If a field has an `enum`, should we require that it be expressed as a codelist?

We can offer tool support for this: both for validating extensions, and for automatically putting the codes from the codelists into the `enum`. (Both tools now exist.)

http://standard.open-contracting.org/latest/en/schema/conformance_and_extensions/#extensions

**Comment by jpmckinney** [comment]
Possible new requirements (related to https://github.com/open-contracting/standard-maintenance-scripts/issues/48):

* Conformant extensions should not delete fields from the standard's schema.
* Conformant extensions should not change the properties of fields from the standard's schema. Note: api_extension changes `required` to `[]` and ocds_milestone_documents_extension changes `deprecated` to `null`.

This is to avoid an extension changing the semantics of the standard's schema and thereby breaking interoperability.

If an extension desires to further document the usage of a field, it should do so through documentation, rather than through changing a field's `description` property.

**Comment by jpmckinney** [comment]
Possible new requirements (related https://github.com/open-contracting/standard-maintenance-scripts/issues/45):

* Definition names should be UpperCamelCase.
* Field names should be lowerCamelCase.
* Both names should contain only ASCII alphabetical letters.
* If an acronym is used within a name, the acronym should be all uppercase, unless it is at the beginning of the name, in which case it should be all lowercase.

**Comment by duncandewhurst** [comment]
I'm in favour of all of the proposed requirements. However, there is some cross-over with parts of the [schema style guide](https://ocds-standard-development-handbook.readthedocs.io/en/latest/meta/schema_style_guide.html#), so we should consider how to keep those in sync, or whether the style guide should defer to the extensions conformance section for some points.

> If a field has an `enum`, should we require that it be expressed as a codelist?

I think so, so that titles and descriptions for the enum values can be translated.

**Comment by jpmckinney** [comment]
Yes, once this documentation exists, we can remove some content from the style guide and link to these docs.

This is quite an old issue. Some additional rules for extensions are expressed by the schema files for CSV codelists and extension.json files: https://github.com/open-contracting/standard-maintenance-scripts/tree/main/schema

These [tests](https://github.com/open-contracting/standard-maintenance-scripts/tree/main/tests) might also go beyond the above. (The [jscc](https://jscc.readthedocs.io/en/latest/api/testing/checks.html) documentation might be easier to read.)

I'd consider this issue to be lower priority, as we have internal tools for checking extensions. For the rules to be really useful, we'd have to consider an online tool – but there are not a huge number of unregistered extensions.

**Comment by jpmckinney** [comment]
See also #1571 on rules for publishing extensions.

**Comment by jpmckinney** [comment]
Noting that `normative.md` mentions `extension-schema.json` and `codelist-schema.json`. If this issue mentions those, we should move them to this repository, and update `normative.md`.

**Comment by duncandewhurst** [comment]
@jpmckinney I know you said this was lower priority, but given that most other issues are blocked or pending review, shall @odscjen or I prepare a PR to update the extensions conformance documentation?

**Comment by jpmckinney** [comment]
Sure. Let's keep it succinct (e.g. bullet form, with subheadings to organize content as needed). Once the PR is approved, there can be a corresponding PR to delete content from the handbook and reference this page instead (it can link to the staging page until 1.2 is released).

**Comment by duncandewhurst** [comment]
> * Conformant extensions should not change the properties of fields from the standard's schema. Note: api_extension changes `required` to `[]` and ocds_milestone_documents_extension changes `deprecated` to `null`.

@jpmckinney does this rule need to be refined to allow an extension to:

* Make an existing optional field `required`?
* Impose a greater `minLength` on an existing field?

I think the idea of allowing extensions to impose additional validation constraints on existing fields was discussed in the past, but I don't know if any extensions actually do that.

**Comment by jpmckinney** [comment]
I don't think we need to make any exceptions to the rule (which I think agrees with what you said).

In principle, the schema are intended to describe correct structure, but are not intended to describe correct *data*, except in a few OCDS-specific cases (for example, in principle, there could be a `pattern` validation property on the `ocid` field).

In practice, it is *possible* that a publisher might want to add additional validation properties to better describe correct data – but as you say, that is not a behavior we've observed. Publishers can validate their data however they like, and I think it's okay to treat such JSON Schema as not being extensions.

**Comment by duncandewhurst** [comment]
Interesting, I hadn't thought about that distinction before. I don't see a problem with it, but we might want to document that principle somewhere (if not in the OCDS documentation, then in the handbook).

**Comment by jpmckinney** [comment]
We can maybe add a note at https://ocds-standard-development-handbook.readthedocs.io/en/latest/extensions/index.html
